keps/sig-testing/1746-stage-e2e-test-framework/README.md

```diff
@@ -0,0 +1,610 @@
+rewrite e2e test framework functions using official packages and APIs to match
+the current desired behaviour.
+For the most part this should be straightforward and it will only cause issues
+when one of the other two following conditions are present.
+
+Bullet point number 2 is present in cases where e2e tests involve collecting
+metrics from unpublished APIs.
+As an example, see
+[kubeletstatsv1alpha1](https://github.com/kubernetes/kubernetes/blob/ba3bf32300574d86c98657981c96ca609de787a2/test/e2e/framework/resource_usage_gatherer.go#L37)
+being imported to collect Kubelet statistics.
+We want to prevent this kind of internal dependencies in order to ensure that
+all artifacts needed for e2e testing are properly surfaced to users.
+To deal with these cases we will collaborate with the owning SIGs to have the
+necessary internal APIs used in e2e tests published in
+https://github.com/kubernetes/api.
+
+Bullet point 3 will be dealt in a similar manner by working with owning SIGs to
+export any constants and variables defined in core Kubernetes that are required for e2e tests.
+
+### Managing Test Packages
+
+Decoupling the e2e test framework from other test packages will proceed in a
+similar manner.
+
+We will take a look at the utility functions surfaced by the e2e test framework
+and we will see how and why they are used in e2e tests.
+If we encounter utilities that are used sparingly in only a single test package
+then we will opt to move it to that test package.
+
+On the other hand, if we see a lot of dependencies and complexity introduced by
+requiring other test packages then we will work to migrate the essential parts
+within the e2e test framework.
+
+Ultimately, we want a library that is well encapsulated and has useful methods
+for maintainers to write e2e tests.
+
+
+### Notes
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Kubemark
+
+One of the important e2e test components that we have been looking at is
+[Kubemark](https://github.com/kubernetes/kubernetes/tree/72cdc8c2112cf14daeb8b4f29cb524c3aba890b4/test/kubemark)
+which brings in a significant portion of dependencies from core Kubernetes.
+Some of the use of internal unpublished APIs is believed to be related to this
+and some other components built for scalability testing.
+
+Given that Kubemark is not part of the e2e test framework, the current plan is
+that no particular amount of work, other than the already mentioned, will be
+needed to decouple the e2e test framework from Kubemark.
+
+There is also the possibility that if Kubemark is the only reason why internal
+Kubernetes APIs (such as the Kubelet stat API) are needed that we move the codes
+that makes use of this to kubemark itself.
+
+
+
+### Risks and Mitigations
```

Comment:
There is also a risk that we are unable to make changes we need for testing kubernetes in a timely fashion due to concerns about breaking external consumers, and a risk that we are unable to test internal aspects of Kubernetes.

Reply:
Building on my line of reasoning from https://github.com/kubernetes/enhancements/pull/1763#discussion_r424736251
I don't think there is a risk that we will be unable to test internal aspects of Kubernetes because code for e2e tests can still be incorporated within core Kubernetes and used in Kubernetes e2e tests.

As to the former point, testing Kubernetes in a timely fashion could be done by adding the code in the test directory as it is done now.
Utility functions of the test framework are for the most part commodities, not necessities.